<a href="https://colab.research.google.com/github/ValeriiLubiako/MyFirstRepository/blob/master/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22Untitled5_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Основы языка Python для аналитиков (семинары)#
##Урок 3. Изменение таблиц в Pandas##

###Условие 1: Задача 1###
_Скачать данные c домами._

_Считать данные с помощью pandas_

__Вывести на экран первые 5 строк__

1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю <br>
стоимость за один кв. метр общей площади

1.2 Создать новый признак delta_renovated, который будет содержать разницу в <br>годах между годом реновацией дома и годом постройки дома. Если реновации <br> дома не было, то в новом признаке поставьте 0.

1.3 Создайте признаки года продажи, месяца продажи

1.4 Удалите признаки date, zipcode, lat, long

In [4]:
import pandas as pd

In [2]:
from google.colab import files
uploaded=files.upload()

Saving kc_house_data.csv to kc_house_data.csv


In [5]:
df=pd.read_csv('kc_house_data.csv', encoding='latin-1')

Вывести на экран первые 5 строк

In [6]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

1.1 Создать новый признак price_per_sq_lot, который будет содержать среднюю
стоимость за один кв. метр общей площади

In [9]:
df['price_per_sq_lot'] = df['price']/df['sqft_lot']

In [10]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  price_per_sq_lot  
0        5650         39.274336  
1        7639         74.288870  
2        8062         18.000000  
3        5000        120.800000  
4        7503         63.118812  

[5 rows x 22 columns]

1.2. Создать новый признак delta_renovated, который будет содержать разницу в
годах между<br> годом реновацией дома и годом постройки дома.<br> Если реновации
дома не было, то в новом признаке поставьте 0.

In [17]:
import numpy as np

In [19]:
df['delta_renovated'] = np.where(df['yr_renovated'] == 0, 0, df['yr_renovated']-df['yr_built'])

In [20]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_basement  yr_built  \
0      5650     1.0           0     0  ...              0      1955   
1      7242     2.0           0     0  ...            400      1951   
2     10000     1.0           0     0  ...              0      1933   
3      5000     1.0           0     0  ...            910      1965   
4      8080     1.0           0     0  ...              0      1987   

   yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  \
0             0    98178  47.5112 -122.257           1340        5650   
1          1991    98125  47.7210 -122.319           1690        7639   
2             0    98028  47.7379 -122.233           2720        8062   
3             0    98136  47.5208 -122.393           1360        5000   
4             0    98074  47.6168 -122.045           1800        7503   

   price_per_sq_lot  delta_renovated  
0         39.274336                0  
1         74.288870               40  
2         18.000000                0  
3        120.800000                0  
4         63.118812                0  

[5 rows x 23 columns]

1.3 Создайте признаки года продажи, месяца продажи

In [26]:
df['yr_sold'] = df['date'][0:3]

In [27]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  zipcode  \
0      5650     1.0           0     0  ...      1955             0    98178   
1      7242     2.0           0     0  ...      1951          1991    98125   
2     10000     1.0           0     0  ...      1933             0    98028   
3      5000     1.0           0     0  ...      1965             0    98136   
4      8080     1.0           0     0  ...      1987             0    98074   

       lat     long  sqft_living15  sqft_lot15  price_per_sq_lot  \
0  47.5112 -122.257           1340        5650         39.274336   
1  47.7210 -122.319           1690        7639         74.288870   
2  47.7379 -122.233           2720        8062         18.000000   
3  47.5208 -122.393           1360        5000        120.800000   
4  47.6168 -122.045           1800        7503         63.118812   

   delta_renovated          yr_sold  
0                0  20141013T000000  
1               40  20141209T000000  
2                0  20150225T000000  
3                0              NaN  
4                0              NaN  

[5 rows x 24 columns]

In [35]:
def partline(my_str:str):
  my_str = my_str[0:4]
  return my_str

In [41]:
df['yr_sold'] = df['date'].apply(partline)

In [42]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  zipcode  \
0      5650     1.0           0     0  ...      1955             0    98178   
1      7242     2.0           0     0  ...      1951          1991    98125   
2     10000     1.0           0     0  ...      1933             0    98028   
3      5000     1.0           0     0  ...      1965             0    98136   
4      8080     1.0           0     0  ...      1987             0    98074   

       lat     long  sqft_living15  sqft_lot15  price_per_sq_lot  \
0  47.5112 -122.257           1340        5650         39.274336   
1  47.7210 -122.319           1690        7639         74.288870   
2  47.7379 -122.233           2720        8062         18.000000   
3  47.5208 -122.393           1360        5000        120.800000   
4  47.6168 -122.045           1800        7503         63.118812   

   delta_renovated  yr_sold  
0                0     2014  
1               40     2014  
2                0     2015  
3                0     2014  
4                0     2015  

[5 rows x 24 columns]

In [38]:
def partline46(my_str:str):
  my_str = my_str[4:6]
  return my_str

In [43]:
df['month_sold'] = df['date'].apply(partline46)

In [44]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_renovated  zipcode      lat  \
0      5650     1.0           0     0  ...             0    98178  47.5112   
1      7242     2.0           0     0  ...          1991    98125  47.7210   
2     10000     1.0           0     0  ...             0    98028  47.7379   
3      5000     1.0           0     0  ...             0    98136  47.5208   
4      8080     1.0           0     0  ...             0    98074  47.6168   

      long  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0 -122.257           1340        5650         39.274336                0   
1 -122.319           1690        7639         74.288870               40   
2 -122.233           2720        8062         18.000000                0   
3 -122.393           1360        5000        120.800000                0   
4 -122.045           1800        7503         63.118812                0   

   yr_sold  month_sold  
0     2014          10  
1     2014          12  
2     2015          02  
3     2014          12  
4     2015          02  

[5 rows x 25 columns]

1.4 Удалите признаки date, zipcode, lat, long

In [45]:
df.drop(columns=['date', 'zipcode', 'lat',  'long'], inplace = True)



In [46]:
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  ...  sqft_above  sqft_basement  yr_built  \
0           0     0          3  ...        1180              0      1955   
1           0     0          3  ...        2170            400      1951   
2           0     0          3  ...         770              0      1933   
3           0     0          5  ...        1050            910      1965   
4           0     0          3  ...        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  price_per_sq_lot  delta_renovated  \
0             0           1340        5650         39.274336                0   
1          1991           1690        7639         74.288870               40   
2             0           2720        8062         18.000000                0   
3             0           1360        5000        120.800000                0   
4             0           1800        7503         63.118812                0   

   yr_sold month_sold  
0     2014         10  
1     2014         12  
2     2015         02  
3     2014         12  
4     2015         02  

[5 rows x 21 columns]

##Условие 2: 2 задача##
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})

2.1 Присоедините к таблице clients данные по домам через метод join

2.2 Присоедините к таблице clients данные по домам через метод merge

Это нужно, чтобы понимать, какие дома покупались клиентами
house_id - это индексы датафрейма с домами